### PR TITLE
Allow user to control if they are signed out of all devices when changing password

### PR DESCRIPTION
**Pull request by hughns**
Depends on https://github.com/matrix-org/matrix-js-sdk/pull/2285

Attempts to resolve https://github.com/vector-im/element-web/issues/2671 for Element Web.

Forgot password flow
---

If the HS does have the capability (supports r0.6.1) then we present the option of whether to sign out other devices:

image

And if the checkbox is ticked then we show a warning/confirmation modal:

image

If the HS does not have the capability then the behaviour is as it is now, but with a revised wording in the confirmation modal:

image

image

Change password from Settings
---

image

If the HS does have the capability or the user only has one device (the current one) then we proceed immediately to changing the password without signing out other devices:

image

Otherwise, we continue to show a warning/confirmation modal, but with some revised wording:

image

image

Notes:
Make sign out on all devices during password reset optional. Changing password in Settings does not sign out all devices.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Allow user to control if they are signed out of all devices when changing password ([\#8259](https://github.com/matrix-org/matrix-react-sdk/pull/8259)). Fixes vector-im/element-web#2671. Contributed by @hughns.

Preview: https://pr8259--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by hughns** [comment]
I'm not sure if this is considered a breaking change or not: the `confirm` property in the `ChangePassword` component has been removed.

If necessary I can reinstate `confirm` and if `confirm=true` then show the existing modal and actually sign out all devices.

**Comment by hughns** [comment]
I've put a check box on the password reset form:

image

Only if the box is checked then a revised confirmation modal is shown:

image

File: src/components/structures/auth/ForgotPassword.tsx [hunk 1]
```diff
@@ -72,6 +73,8 @@ interface IState {
     serverDeadError: string;
 
     currentHttpRequest?: Promise<any>;
+
+    logout_devices: boolean;
```
Comment (t3chguy) [inline src/components/structures/auth/ForgotPassword.tsx:77]: camelCase please

File: src/components/structures/auth/ForgotPassword.tsx [hunk 2]
```diff
@@ -174,24 +178,28 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
...
+                    <div>
+                        { _t(
+                            "Signing out all sessions will reset the end-to-end encryption keys " +
+                            "making encrypted chat history unreadable. Set up " +
+                            "Key Backup or export your room keys from another session before resetting your " +
+                            "password.",
+                        ) }
+                    </div>,
+                button: _t('Continue'),
+                onFinished: (confirmed) => {
```
Comment (t3chguy) [inline src/components/structures/auth/ForgotPassword.tsx:194]:
```suggestion
                onFinished: (confirmed: boolean) => {
```

**Review by t3chguy — APPROVED** [review]
Looks sane to me otherwise!

File: src/components/structures/auth/ForgotPassword.tsx [hunk 3]
```diff
@@ -174,24 +178,28 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
...
+        if (this.state.logout_devices) {
+            Modal.createTrackedDialog('Forgot Password Warning', '', QuestionDialog, {
+                title: _t('Warning!'),
+                description:
+                    <div>
+                        { _t(
+                            "Signing out all sessions will reset the end-to-end encryption keys " +
+                            "making encrypted chat history unreadable. Set up " +
+                            "Key Backup or export your room keys from another session before resetting your " +
+                            "password.",
```
Comment (dkasak) [inline src/components/structures/auth/ForgotPassword.tsx:190]: Could we use this opportunity to change this message so it's less misleading? It talks about *resetting* end-to-end encryption keys, which makes it sound like something will happen server-side regarding this.

But in truth, it's the clients willingly *deleting* their own keys once they notice they've been signed out. So perhaps something like:

```suggestion
                            "Signing out your sessions will delete the message encryption keys stored on " +
                            "them, making encrypted chat history unreadable. Set up Key Backup or export " +
                            "your message keys from one of your sessions before proceeding.",
```

(Aside: also, "sessions" vs "devices"? Did we settle on one of those? I feel like we're still using them interchangeably all over the place.)

File: src/i18n/strings/en_EN.json [hunk 4]
```diff
@@ -3137,11 +3136,12 @@
     "Really reset verification keys?": "Really reset verification keys?",
     "Skip verification for now": "Skip verification for now",
     "Failed to send email": "Failed to send email",
-    "Changing your password will reset any end-to-end encryption keys on all of your sessions, making encrypted chat history unreadable. Set up Key Backup or export your room keys from another session before resetting your password.": "Changing your password will reset any end-to-end encryption keys on all of your sessions, making encrypted chat history unreadable. Set up Key Backup or export your room keys from another session before resetting your password.",
+    "Signing out all sessions will reset the end-to-end encryption keys making encrypted chat history unreadable. Set up Key Backup or export your room keys from another session before resetting your password.": "Signing out all sessions will reset the end-to-end encryption keys making encrypted chat history unreadable. Set up Key Backup or export your room keys from another session before resetting your password.",
```
Comment (dkasak) [inline src/i18n/strings/en_EN.json:3139]: And of course this needs changing too if we go [with my suggestion above](https://github.com/matrix-org/matrix-react-sdk/pull/8259/files#r846011928).

File: src/components/structures/auth/ForgotPassword.tsx (hunk 3 above)
Comment (duxovni) [reply src/components/structures/auth/ForgotPassword.tsx:190]: The last time @amshakal and I talked about it for the device list view, we decided "devices" was the least confusing word to use, but I don't think there's broad consensus on this beyond the Crypto Team.

Comment (dkasak) [reply src/components/structures/auth/ForgotPassword.tsx:190]: FWIW, that's my take on this as well.

**Comment by hughns** [comment]
I've incorporated the wording suggestions into the latest commits.

The screens now look like this:

image

image

**Comment by Mikaela** [comment]
Will it really sign out all devices or all other devices (excluding this one)?

**Comment by hughns** [comment]
> Will it really sign out all devices or all other devices (excluding this one)?

My thinking was: the current device isn't signed in at the moment (otherwise the user wouldn't need to reset the password).

However, I think either wording is valid:

- "Sign out all devices"

- "Sign out all other devices"

I would advocate for whichever we think is the clearest.

**Comment by dkasak** [comment]
Note that Element Web also has functionality to change the password while logged in. Does this PR only handle the password resetting case?

File: src/components/structures/auth/ForgotPassword.tsx [hunk 5]
```diff
@@ -314,6 +322,11 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
                         autoComplete="new-password"
                     />
                 </div>
+                <div className="mx_AuthBody_fieldRow">
+                    <StyledCheckbox onChange={() => this.setState({ logoutDevices: !this.state.logoutDevices })} checked={this.state.logoutDevices}>
+                        { _t("Sign out all devices") }
+                    </StyledCheckbox>
```
Comment (t3chguy) [inline src/components/structures/auth/ForgotPassword.tsx:328]: We actually don't know if the server will respect this flag, it may be too old

File: src/components/views/settings/ChangePassword.tsx [hunk 6]
```diff
@@ -140,7 +101,8 @@ export default class ChangePassword extends React.Component<IProps, IState> {
             phase: Phase.Uploading,
         });
 
-        cli.setPassword(authDict, newPassword).then(() => {
+        // This no longer logs out all sessions:
+        cli.setPassword(authDict, newPassword, false).then(() => {
```
Comment (t3chguy) [inline src/components/views/settings/ChangePassword.tsx:105]: We can't assume this actually, we do not know if the connected server supports this API yet

**Review by t3chguy — CHANGES_REQUESTED** [review]
We need a way of checking if the server will respect the flag, otherwise we're misleading the user in both cases (forgot and change password)

**Comment by hughns** [comment]
> We need a way of checking if the server will respect the flag, otherwise we're misleading the user in both cases (forgot and change password)

This is a very good point.

**Comment by clokep** [comment]
Just an FYI that Synapse has supported this since v1.12.0 released on 2020-03-23 and it seems that it was added in r0.6.1 of the spec: https://matrix.org/docs/spec/client_server/r0.6.1.html#changelog

**Comment by t3chguy** [comment]
Sure but other server implementations may not support it, and promising the user something relating to security which doesn't work without warning is baaaad

**Comment by clokep** [comment]
> Sure but other server implementations may not support it, and promising the user something relating to security which doesn't work without warning is baaaad

Of course, my point is that you should be able to key on the r0.6.1 version to know if it is supported!

**Comment by t3chguy** [comment]
> Of course, my point is that you should be able to key on the r0.6.1 version to know if it is supported!

Not really, Dendrite advertises 0.6.1 support but doesn't support this.

```
t3chguy@Michaels-MacBook-Pro-2 ~> curl https://dendrite.matrix.org/_matrix/client/versions --silent | jq .
{
  "versions": [
    "r0.0.1",
    "r0.1.0",
    "r0.2.0",
    "r0.3.0",
    "r0.4.0",
    "r0.5.0",
    "r0.6.1"
  ],
  "unstable_features": {
    "org.matrix.e2e_cross_signing": true,
    "org.matrix.msc2836": true,
    "org.matrix.msc2946": true
  }
}
```

Dendrite just hardcodes the value to true even though it demarshalls it properly - https://github.com/matrix-org/dendrite/blob/f2e550efd832662e6c032bebfef4a68da0b4c8ee/clientapi/routing/password.go#L37

**Comment by hughns** [comment]
In parallel to determining the homeserver capabilities I've worked on a revised UI to take account of this.

For testing in this PR I've tied it to whether you are connecting to matrix.org homeserver vs any other homeserver.

On the wording in the below I wasn't sure about talking about the "homeserver" and getting an upgrade... let me know what you think.

Forgot password
---

If the HS does not have the capability then the behaviour is as it is now, but with a revised wording in the confirmation modal:

image

image

If the HS does have the capability then we present the option:

image

And if the checkbox is ticked then we show a warning/confirmation modal:

image


Settings
---

If the HS does not have the capability then we continue to show a warning/confirmation modal, but with some revised wording:

image

If the HS does have the capability then we proceed immediately to changing the password without signing out other devices.

**Comment by t3chguy** [comment]
> If the HS does not have the capability then the behaviour is as it is now, but with a revised wording in the confirmation modal:

Missing space between sentences 

image

File: src/components/structures/auth/ForgotPassword.tsx (hunk 5 above)
Comment (hughns) [reply src/components/structures/auth/ForgotPassword.tsx]: The UI aspect of this is resolved in the latest changes.

File: src/i18n/strings/en_EN.json [hunk 7]
```diff
@@ -3143,18 +3143,21 @@
...
+    "Resetting your password on this homeserver will cause all of your devices to be signed out. This will delete the message encryption keys stored on them, making encrypted chat history unreadable.": "Resetting your password on this homeserver will cause all of your devices to be signed out. This will delete the message encryption keys stored on them, making encrypted chat history unreadable.",
+    "Signing out your devices will delete the message encryption keys stored on them, making encrypted chat history unreadable.": "Signing out your devices will delete the message encryption keys stored on them, making encrypted chat history unreadable.",
+    "If you want to access your encrypted chat history then set up Key Backup or export your message keys from one of your other devices before proceeding.": "If you want to access your encrypted chat history then set up Key Backup or export your message keys from one of your other devices before proceeding.",
```
Comment (dkasak) [inline src/i18n/strings/en_EN.json:3148]: Upon re-reading this, I think a better wording would be:

> If you want to retain access to your chat history in encrypted rooms, set up Key Backup or export your message keys from one of your other devices before proceeding.

**Comment by dkasak** [comment]
image

The wall of text in the warning would be much more readable if broken up into a separate paragraph as marked. So the "If you want to..." sentence goes into a new paragraph.

**Comment by hughns** [comment]
Latest iteration:

Forgot password
---

If the HS does not have the capability then the behaviour is as it is now, but with a revised wording in the confirmation modal:

image

image

If the HS does have the capability then we present the option:

image

And if the checkbox is ticked then we show a warning/confirmation modal:

image


Settings
---

If the HS does not have the capability then we continue to show a warning/confirmation modal, but with some revised wording:

image

If the HS does have the capability then we proceed immediately to changing the password without signing out other devices.

**Comment by hughns** [comment]
> A HS agnostic approach could be to use [Capability negotiation](https://spec.matrix.org/v1.2/client-server-api/#mchange_password-capability) and extend the [m.change_password capability](https://spec.matrix.org/v1.2/client-server-api/#mchange_password-capability) to say if this is supported or not.

I started implementing this as a Capability but it became apparent that `GET /_matrix/client/v3/capabilities` requires the caller to be authenticated which doesn't help us during the forgotten password flow.

Is the `versions` portion of `GET /_matrix/client/versions` as @clokep suggested perhaps the better direction to go?

Any other suggestions?

**Comment by t3chguy** [comment]
> It looks like we are checking to see if the email was verified by attempting to change the password, assuming we can check the verification by some other API call (like a call to /_matrix/client/v3/capabilities) this could work.

This is a common issue with UIA (User Interactive Auth) in Matrix, there's no way to "test" whether its ready/successful without executing the action.

File: src/i18n/strings/en_EN.json (hunk 7 above)
Comment (hughns) [reply src/i18n/strings/en_EN.json]: Agreed. Wording updated.

**Comment by dkasak** [comment]
> If we are doing that I would be in favour of not even asking for the new password until the email confirmation has been done and so on the initial screen the use only enters their email, then once verified they see a second screen where they enter their new password and choose whether to logout devices.

Just for reference purposes: You've stumbled on to a [long-standing issue
](https://github.com/vector-im/element-web/issues/18743) (spun out from this [mega-issue](https://github.com/vector-im/element-web/issues/2780)). I agree that this absolutely needs to change.

**Comment by clokep** [comment]
> > A HS agnostic approach could be to use [Capability negotiation](https://spec.matrix.org/v1.2/client-server-api/#mchange_password-capability) and extend the [m.change_password capability](https://spec.matrix.org/v1.2/client-server-api/#mchange_password-capability) to say if this is supported or not.
> 
> I started implementing this as a Capability but it became apparent that `GET /_matrix/client/v3/capabilities` requires the caller to be authenticated which doesn't help us during the forgotten password flow.
> 
> Is the `versions` portion of `GET /_matrix/client/versions` as @clokep suggested perhaps the better direction to go?
> 
> Any other suggestions?

Unfortunately I think adding a capability or to versions would require a spec change here, which is not great.

I see it as OK to check if the homeserver supports `r0.6.1` or later versions of the spec. Since Dendrite will logout all devices in that case (until the bug is fixed) it shouldn't cause any security issue, it just falls back to the current behavior and will feel slightly buggy.

**Comment by t3chguy** [comment]
This should have been covered by the MSC yet wasn't
The MSC seemingly landed with only partial implementation (server, no client)

I think this is a failure of the MSC process

**Comment by t3chguy** [comment]
> it shouldn't cause any security issue

In the case of pw reset this results in your other sessions getting nuked and if you don't have key backup (and no warning due to the client thinking the server supports it) can cause irreversible e2ee data loss.

**Comment by t3chguy** [comment]
Upon re-read, https://github.com/matrix-org/dendrite/blob/f2e550efd832662e6c032bebfef4a68da0b4c8ee/clientapi/routing/password.go#L37 isn't forced to true, but rather the default value. Dendrite supports this field.

**Comment by t3chguy** [comment]
And Conduit does not advertise 0.6.1 support, only 0.6.0 - so using 0.6.1 should be ok

File: src/components/views/settings/ChangePassword.tsx [hunk 8]
```diff
@@ -82,48 +82,64 @@ export default class ChangePassword extends React.Component<IProps, IState> {
...
 
-    private onChangePassword(oldPassword: string, newPassword: string): void {
+    private async onChangePassword(oldPassword: string, newPassword: string): Promise<void> {
         const cli = MatrixClientPeg.get();
 
-        if (!this.props.confirm) {
-            this.changePassword(cli, oldPassword, newPassword);
-            return;
-        }
+        const serverSupportsControlOfDevicesLogout: boolean = await cli.doesServerSupportLogoutDevices();
```
Comment (t3chguy) [inline src/components/views/settings/ChangePassword.tsx:88]:
```suggestion
        const serverSupportsControlOfDevicesLogout = await cli.doesServerSupportLogoutDevices();
```

We prefer implicit types where the right side of the assignment is well-typed

File: src/components/views/settings/ChangePassword.tsx (hunk 6 above)
Comment (hughns) [reply src/components/views/settings/ChangePassword.tsx]: This is now dealt with by querying the spec versions that the server supports.

File: src/components/views/settings/ChangePassword.tsx [hunk 9]
```diff
@@ -82,48 +82,64 @@ export default class ChangePassword extends React.Component<IProps, IState> {
...
-            onFinished: (confirmed) => {
-                if (confirmed) {
-                    this.changePassword(cli, oldPassword, newPassword);
-                }
-            },
-        });
+        if (serverSupportsControlOfDevicesLogout) {
+            // don't log user out of all devices
+            this.changePassword(cli, oldPassword, newPassword, /* logoutDevices = */ false);
+        } else {
```
Comment (t3chguy) [inline src/components/views/settings/ChangePassword.tsx:93]: This block (won't let me select to the other end of the else) is a pain to read, Modal.createTrackedDialog has an async signature which you can use to DRY this area.

e.g

```javascript
const logoutDevices = serverSupportsControlOfDevicesLogout ? false : undefined;

if (!serverSupportsControlOfDevicesLogout && this.props.confirm) {
    const { finished } = Modal.createTrackedDialog<[boolean]>(...);
    const [confirmed] = await finished;
    if (!confirmed) return;
}

this.changePassword(cli, oldPassword, newPassword, logoutDevices);
```

or similar

File: src/components/structures/auth/ForgotPassword.tsx [hunk 10]
```diff
@@ -174,24 +198,37 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
...
+                onFinished: (confirmed: boolean) => {
+                    if (confirmed) {
+                        this.submitPasswordReset(this.state.email, this.state.password, this.state.logoutDevices);
+                    }
+                },
+            });
+        } else {
+            this.submitPasswordReset(this.state.email, this.state.password, this.state.logoutDevices);
+        }
     };
```
Comment (t3chguy) [inline src/components/structures/auth/ForgotPassword.tsx:232]: We can use a similar async signature here as in the other comment to DRY

File: src/components/views/settings/ChangePassword.tsx (hunk 9 above)
Comment (hughns) [reply src/components/views/settings/ChangePassword.tsx]: Much nicer. Thanks!

**Review by robintown — COMMENTED** [review]
Using this to change your password in settings results in a dialog that says "You will not receive push notifications on other sessions until you log back in to them", which I believe is no longer accurate

**Review by t3chguy — APPROVED** [review]

**Comment by hughns** [comment]
> Using this to change your password in settings results in a dialog that says "You will not receive push notifications on other sessions until you log back in to them", which I believe is no longer accurate

Good spot.

I've pushed changes that take account of whether all devices were signed out during the password change.

If the server supports not signing out all devices:

image

Otherwise:

image

**Comment by hughns** [comment]
Whilst making these changes I realised that in the case the user only has one logged in device we could simplify the wording throughout. i.e. we don't need to warn them when there aren't any other devices that would actually be signed out.

I don't think it would be hard to roll into this change. Good idea? Objections?

**Review by t3chguy — APPROVED** [review]

**Review by nadonomy — APPROVED** [review]
@hughns as functional changes these LGTM, so approving on behalf of design.

There's a lot we could improve in the UX copy (too much jargon, not layperson friendly) but a lot of that copy is already on production predating this PR.

Anticipating we'll improve this in earnest when we make larger changes to reset password (OIDC/auth?).
